### PR TITLE
Add explicit check for modifying data in delta table with CDF enabled

### PR DESCRIPTION
**Pull request by homar**
## Description
Add explicit check for modifying data in delta table with CDF enabled.
Inserts could be allowed as they don't have to do anything special to support CDF but they are blocked by writer version check. 

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Add explicit check for modifying data in delta table with CDF enabled. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
Sent #14508 within the repository.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 1]
```diff
@@ -422,6 +422,14 @@ public static Map<String, String> getCheckConstraints(MetadataEntry metadataEntr
                 .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
+    public static boolean changeDataFeedEnabled(MetadataEntry metadataEntry)
+    {
+        String enableChangeDataFeed = metadataEntry.getConfiguration().entrySet().stream()
+                .filter(entry -> entry.getKey().equals("delta.enableChangeDataFeed"))
+                .findAny().map(Map.Entry::getValue).orElse("false");
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:429]: nit: How about using `getOrDefault`?
```java
String enableChangeDataFeed = metadataEntry.getConfiguration().getOrDefault("delta.enableChangeDataFeed", "false");
```

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 1 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java]: I was not aware of this. thanks

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 2]
```diff
@@ -422,6 +422,12 @@ public static Map<String, String> getCheckConstraints(MetadataEntry metadataEntr
                 .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
+    public static boolean changeDataFeedEnabled(MetadataEntry metadataEntry)
+    {
+        String enableChangeDataFeed = metadataEntry.getConfiguration().getOrDefault("delta.enableChangeDataFeed", "false");
+        return enableChangeDataFeed.equalsIgnoreCase("true");
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:428]: Is `java.lang.Boolean#parseBoolean` what Delta is using?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java [hunk 3]
```diff
@@ -516,4 +516,29 @@ public void testMetadataOperationsRetainCheckConstraints()
...
+    {
+        String tableName = "test_writes_into_table_with_CDF_" + randomTableSuffix();
+        try {
+            onDelta().executeQuery("CREATE TABLE default." + tableName + " (a INT, b INT) " +
+                    "USING DELTA " +
+                    "LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "'" +
+                    "TBLPROPERTIES (delta.enableChangeDataFeed = true)");
+
+            assertQueryFailure(() -> onTrino().executeQuery("INSERT INTO delta.default." + tableName + " VALUES (1, 2)"))
+                    .hasMessageMatching(".* Table .* requires Delta Lake writer version 4 which is not supported");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksInsertCompatibility.java:531]: So it was blocked before the change as well and the change just introduces an explicit check with an explicit message.

Let's have a different commit and PR title then

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 2 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:428]: This is Delta implementation:
```scala
  val CHANGE_DATA_FEED = buildConfig[Boolean](
    "enableChangeDataFeed",
    "false",
    _.toBoolean,
    _ => true,
    "needs to be a boolean.",
    alternateConfs = Seq(CHANGE_DATA_FEED_LEGACY))

  def toBoolean: Boolean = parseBoolean(toString)

  private def parseBoolean(s: String): Boolean =
    if (s != null) s.toLowerCase match {
      case "true" => true
      case "false" => false
      case _ => throw new IllegalArgumentException("For input string: \""+s+"\"")
    }
    else
      throw new IllegalArgumentException("For input string: \"null\"")
```

**Comment by homar** [comment]
maven checks were killed because they were super slow, doesn't look like anything actionable

**Comment by ebyhr** [comment]
Could you rebase on upstream to resolve conflicts?

**Comment by colebow** [comment]
Could you clarify what the user impact of this is for the release note?

It seems to me like this would always fail (due to that writer version check), but now it's failing with better error messaging?

**Comment by ebyhr** [comment]
> It seems to me like this would always fail (due to that writer version check), but now it's failing with better error messaging?

Correct. There's no need to mention in a release note.

**Pull request merged**